bot/utils/services.py

```diff
@@ -1,50 +1,106 @@
+import dataclasses
+import datetime
+import json
+import typing
 from typing import Optional
 
+import aiohttp
 from aiohttp import ClientConnectorError
+from async_rediscache import RedisCache
 
 import bot
 from bot.constants import URLs
 from bot.log import get_logger
 
 log = get_logger(__name__)
+T = typing.TypeVar("T")
 
 FAILED_REQUEST_ATTEMPTS = 3
 
+UNFURL_CACHE = RedisCache(namespace="UnfurledRedirects")
```

Comment:
The persistence is useful if we get hit with a raid and suddenly have to unfurl the same URL 300 times a minute.

It's cleared if it's pulled after expiring, which in practice means the data is only ever a day old, or however long it is.

The case is for constants, I think we use it like this elsewhere too.

Reply:
The cache is useful in case of a raid, not necessarily the persistence. For the cache you could use a simple LRU cache as mentioned.

> It's cleared if it's pulled after expiring
Right, but I argue that many URLs are unlikely to be pulled after expiring.

Reply:
I don’t think there’s much harm to doing it in redis. It’s pretty easy to use. As for the point about expiring, the goal isn’t to remove the entries from the table. The goal is to not provide the user with outdated data.

Reply:
In fact it's not a matter of how many URLs won't be pulled, as long as there are any URLs which won't be pulled. The cache will just continue growing indefinitely.

Reply:
That’s true, I’m not sure it’s a problem. We can add an occasional vacuum task, but I don’t know when I can code again

Reply:
> I don’t think there’s much harm to doing it in redis

The harm is that I think it doesn't help much while complicating the code. When you need to manually manage insertions and deletions, make sure it's occasionally vacuumed, and write separate tests for it,  just slapping an LRU cache decorator on  `unfurl_url` seems appealing to me.

Reply:
Can an LRU cache be skipped actually?

Anyways, the reason the worker doesn't handle the cache is because there are much more strict limits on storage in CF workers, and we'd waste a bunch of time waiting for the request + lookup instead of just lookup. It could in theory be done though.

Reply:
It's actually possible to make the distinction. When the unfurl succeeds or when there are too many redirects, return as normal. When there is an error, raise an actual error and the LRU cache won't be able to cache the result.

That said one can argue that we actually do want to cache failures because if we're bombarded with a redirect in a short amount of time, and the first try fails, it's reasonable to assume the rest will too. It's an unlikely scenario, but then again that's one of the arguments for using a cache here in the first place.

Reply:
If an error is raised for all failures, then it's impossible to get partially successful results.

Reply:
Well, I suppose the object could be an attribute of a custom exception, but I'm not sure if this is a nice design in this case.

Reply:
I'm looking at `unfurl_url` and I'm not sure what you mean. The only scenario I see where we might want to not cache is for a 400. For other cases I think it can be assumed that the redirect chain is unlikely to change in the near future (in which the cached result will be relevant). And in any case, the failure reason can be the exception message and retrieved in the `except` clause of the caller.

Reply:
I would still like the bypass_cache argument to be a thing, is that possible with an LRU cache?

Reply:
To expand on that: a malicious attacker can bypass our filters by sending the link once to a safe destination, then changing the destination.

Reply:
This seems to undermine the very idea of any cache, including the redis one. How will you tell when to use it without actually unfurling the link again?

Reply:
You wouldn’t, but you can force update the link if you know the link needs updating. Keep in mind that such an attack would be done by someone that has to be familiar with our systems, so we’ll likely notice that something is wrong if we see someone suddenly posting a ton of links to a seemingly safe destination.

Reply:
You mean with a command? I guess if you want to do it manually you can invalidate the cache entry directly. Looking at [utils/caching.py](https://github.com/python-discord/bot/blob/main/bot/utils/caching.py#L6) you might want to allow caching with no size constraint, similarly to https://docs.python.org/3/library/functools.html#functools.cache

Reply:
The behavior I'm describing is already possible with the current unfurl command added by this PR. It's facilitated by that option.

Reply:
Oh, I didn't review the command yet. Yeah I guess you can just remove the key from the dict first.

Reply:
I discussed this a little more with other core devs, and we settled somewhere on trying to get TTL working with redis. SebastiaanZ/async-rediscache#17 came as a result of that, but after some more discussion with Sebastiaan, it's unlikely to be implemented in rediscache. Regardless, we have a backup plan that'll come as a result of some planned changes which are not yet public. Seeing as I can not currently work on this, either way, I'm going to be holding off until that change is ready.